x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+                super.name(),
+                jwsSignatureAlgorithm,
+                this.allowedSignatureAlgorithms
+            );
+
+            // 2. Skip JWT if issuer does not match the issuer allowed by this realm.
+            final String jwtIssuer = jwtClaimsSet.getIssuer();
+            if ((jwtIssuer == null) || (this.allowedIssuer.contains(jwtIssuer) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow issuer [%s]. Allowed issuer is [%s].",
+                    super.name(),
+                    jwtIssuer,
+                    this.allowedIssuer
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug("Realm [%s] allows issuer [%s]. Allowed issuer is [%s].", super.name(), jwtIssuer, this.allowedIssuer);
+
+            // 3. Skip JWT if audience does not match any of the audiences allowed by this realm.
+            final List<String> jwtAudiences = jwtClaimsSet.getAudience();
+            if ((jwtAudiences == null) || (this.allowedAudiences.stream().anyMatch(jwtAudiences::contains) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow audiences %s. Allowed audiences are %s.",
+                    super.name(),
+                    jwtAudiences,
+                    this.allowedAudiences
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug("Realm [%s] allows audiences %s. Allowed audiences are %s.", super.name(), jwtAudiences, this.allowedAudiences);
+
+            // TODO The implementation of JWT authentication will be completed in a later PR
+            // At this point, this is the right realm to do validation. Trigger AuthenticationResult.unsuccessful() if any problems found.
+
+            // Cache lookup
+            // 1. If present in cache, AuthenticationResult.successful().
+
+            // JWT Authentication
+            // 1. Verify signature (HMAC or RSA)
+            // 2. If present verify nfb <= iat.
+            // 3. If present verify nfb <= exp.
+            // 4. If present verify iat <= exp.
+            // 5. If present, verify nfb + allowedClockSkew < now.
+            // 6. If present, verify iat + allowedClockSkew >= now.
+            // 7. If present, verify exp + allowedClockSkew >= now.
+
+            // Client Authorization
+            switch (this.clientAuthorizationType) {
+                case JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_SHARED_SECRET:
+                    if (this.clientAuthorizationSharedSecret.equals(clientAuthorizationSharedSecret) == false) {
```

Comment:
Can we have a specific case (even if it's only a separate log message) for when the client _doesn't_ provide any secret?
That seems useful when debugging failing clients.

Reply:
This is a good idea. I will add it.

Reply:
Done

Reply:
Resolved.